File: sorts/merge/merge.go
package main

import "fmt"

func main() {
	array1 := [7]int{8, 4, 9, 1, 3, 7, 6}
	mergeSort(array1[:])
	fmt.Println(array1)
	array2 := [9]int{9, 8, 7, 6, 5, 4, 3, 2, 1}
	mergeSort(array2[:])
	fmt.Println(array2)
	array3 := [12]int{9, 9, 9, 7, 7, 5, 3, 5, 7, 5, 3, 3}
	mergeSort(array3[:])
	fmt.Println(array3)
	array4 := [7]int{8, 3, 5, 9, 1, 2, 7}
	mergeSort(array4[:])
	fmt.Println(array4)
}

func mergeSort(array []int) {
	arrayCopy := make([]int, len(array))
	sort(array, arrayCopy, 0, len(array)-1)
}

func sort(array []int, arrayCopy []int, lo int, hi int) {
	if hi <= lo {
		return
	}
	mid := lo + (hi-lo)/2
	sort(array, arrayCopy, lo, mid)
	sort(array, arrayCopy, mid+1, hi)
	merge(array, arrayCopy, lo, mid, hi)
}

func merge(array []int, arrayCopy []int, lo int, mid int, hi int) {
	i := lo
	j := mid + 1
	for i, v := range array {
		arrayCopy[i] = v
	}
	for k := lo; k <= hi; k++ {
		if i > mid {
			array[k] = arrayCopy[j]
			j++
		} else if j > hi {
			array[k] = arrayCopy[i]
			i++
		} else if arrayCopy[j] < arrayCopy[i] {
			array[k] = arrayCopy[j]
			j++
		} else {
			array[k] = arrayCopy[i]
			i++
		}
	}
}

File: sorts/merge/Merge.java
public class Merge {

    private static int[] arrayCopy;

    public static void main(String[] args) {
        int[] array1 = {9, 8, 7, 6, 5, 4, 3, 2, 1};
        mergeSort(array1);
        int[] array2 = {8, 4, 9, 1, 3, 7, 6};
        mergeSort(array2);
        int[] array3 = {9, 9, 9, 7, 7, 5, 3, 5, 7, 5, 3, 3};
        mergeSort(array3);
        int[] array4 = {8, 3, 5, 9, 1, 2, 7};
        mergeSort(array4);
    }

    public static void mergeSort(int[] array) {
        arrayCopy = new int[array.length];
        sort(array, 0, array.length - 1);
        for(int i = 0; i < array.length; i++) {
            System.out.print(array[i] + ", ");
        }
        System.out.println();
    }

    public static void sort(int[] array, int lo, int hi) {
        if (hi <= lo) return;
        int mid = lo + (hi - lo) / 2;
        sort(array, lo, mid);
        sort(array, mid + 1, hi);
        merge(array, lo, mid, hi);
    }

    public static void merge(int[] array, int lo, int mid, int hi) {
        int i = lo;
        int j = mid + 1;
        for(int k = lo; k <= hi; k++) {
            arrayCopy[k] = array[k];
        }
        for(int k = lo; k <= hi; k++) {
            if (i > mid) {
                array[k] = arrayCopy[j++];
            } else if (j > hi) {
                array[k] = arrayCopy[i++];
            } else if (arrayCopy[j] < arrayCopy[i]) {
                array[k] = arrayCopy[j++];
            } else {
                array[k] = arrayCopy[i++];
            }
        }
    }
}
